rust/src/checkpoints.rs

```diff
@@ -140,6 +192,153 @@ async fn create_checkpoint(
     Ok(())
 }
 
+async fn flush_delete_files<T: Fn(&(DeltaDataTypeVersion, ObjectMeta)) -> bool>(
+    storage: &dyn StorageBackend,
+    maybe_delete_files: &mut Vec<(DeltaDataTypeVersion, ObjectMeta)>,
+    files_to_delete: &mut Vec<(DeltaDataTypeVersion, ObjectMeta)>,
+    should_delete_file: T,
+) -> Result<i32, DeltaTableError> {
+    if !maybe_delete_files.is_empty() && should_delete_file(maybe_delete_files.last().unwrap()) {
+        files_to_delete.append(maybe_delete_files);
+    }
+
+    let deleted = files_to_delete
+        .iter_mut()
+        .map(|file| async move {
+            match storage.delete_obj(&file.1.path).await {
+                Ok(_) => Ok(1),
+                Err(e) => Err(DeltaTableError::from(e)),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let mut deleted_num = 0;
+    for x in deleted {
+        match x.await {
+            Ok(_) => deleted_num += 1,
+            Err(e) => return Err(e),
+        }
+    }
```

Comment:
perhaps we can avoid the collection here and update the deleted_num count within the first files_to_delete loop?

Reply:
Hmm, tried but got:


```rust
error[E0308]: mismatched types
   --> rust/src/checkpoints.rs:206:48
    |
206 |       files_to_delete.iter_mut().for_each(|file| async move {
    |  ________________________________________________^
207 | |         match storage.delete_obj(&file.1.path).await {
208 | |             Ok(_) => {
209 | |                 deleted_num += 1;
...   |
213 | |         }
214 | |     });
    | |_____^ expected `()`, found opaque type
    |
   ::: /Users/viirya/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/future/mod.rs:61:43
    |
61  |   pub const fn from_generator<T>(gen: T) -> impl Future<Output = T::Return>
    |                                             ------------------------------- the found opaque type
    |
    = note: expected unit type `()`
             found opaque type `impl futures::Future`
```

when I change `map` to `for_each`.

Looks like I cannot return the error from `for_each` for the `Future` from `delete_obj`.